include/seqan3/io/sequence_file/format_sam.hpp

```diff
@@ -0,0 +1,347 @@
+          }
+
+          if ((std::istreambuf_iterator<char>{stream} == std::istreambuf_iterator<char>{}) &&
+              (!stream.eof()))
+          {
+              stream.get(); // triggers error in stream and sets eof
+          }
+
+          // Jump over the new line character in the end
+          detail::consume(stream_view | ranges::view::take_while(is_cntrl));
+    }
+
+    //!\copydoc sequence_file_output_format_concept::write
+    template <typename stream_type,     // constraints checked by file
+              typename seq_type,        // other constraints checked inside function
+              typename id_type,
+              typename qual_type>
+    void write(stream_type                     & stream,
+               sequence_file_output_options const & options,
+               seq_type                       && sequence,
+               id_type                        && id,
+               qual_type                      && qualities)
+    {
+        if constexpr (!(detail::decays_to_ignore_v<seq_type>))
+        {
+            static_assert(std::ranges::ForwardRange<seq_type> && alphabet_concept<value_type_t<seq_type>>,
+          "The sequence must model std::ranges::ForwardRange and its value type must model seqan3::alphabet_concept. ");
+        }
+        ranges::ostreambuf_iterator stream_it{stream};
+        // ID
+        if constexpr (detail::decays_to_ignore_v<id_type>)
+        {
+            throw std::logic_error{"The ID field may not be set to ignore when writing sam files."};
+        }
+        else
+        {
+            if (ranges::empty(id)) //[[unlikely]]
+                throw std::runtime_error{"The ID field may not be empty when writing sam files."};
+
+            ranges::copy(id, stream_it);
+
+            stream_it = '\t';
+            stream_it = '0'; //FLAG
+            stream_it = '\t';
+            stream_it = '*'; //RNAME
+            stream_it = '\t';
+            stream_it = '0'; //POS
+            stream_it = '\t';
+            stream_it = '0'; //MAPQ
+            stream_it = '\t';
+            stream_it = '*'; //CIGAR
+            stream_it = '\t';
+            stream_it = '*'; //RNEXT
+            stream_it = '\t';
+            stream_it = '0'; //PNEXT
+            stream_it = '\t';
+            stream_it = '0'; //TLEN
+            stream_it = '\t';
+        }
+
+        // Sequence
+        if constexpr (detail::decays_to_ignore_v<seq_type>)
+        {
+            throw std::logic_error{"The SEQ and SEQ_QUAL fields may not both be set to ignore when writing sam files."};
+        }
+        else
+        {
+            if (ranges::empty(sequence)) //[[unlikely]]
+                throw std::runtime_error{"The SEQ field may not be empty when writing sam files."};
+            ranges::copy(sequence | view::to_char, stream_it);
+
+        }
+        stream_it = '\t';
+
+        // Quality line
+        if constexpr (detail::decays_to_ignore_v<qual_type>)
```

Comment:
please also remove the constexpr here because qual_type is allowed to be ignored.

Reply:
I am not sure what you mean by allowed to be ignored, I throw an error if it is set to ignore like I do when sequence is set to ignored. In fastaq for both ifs there is a constexpr and then an exception throw. Why is it different here? 
(Plus if I remove it I get a really long error messaged.)